Add nf-test docs for modules (#2059)
nf-core/website: src/content/docs/contributing/modules.md

if you have a module named `build` this can conflict with some pytest internal behaviour. This results in no tests being run (i.e. receiving a message of `collected 0 items`). In this case rename the `tests/<module>/build` directory to `tests/<module>/build_test`, and update the corresponding `test.yml` accordingly. An example can be seen with the [`bowtie2/build` module tests](https://github.com/nf-core/modules/tree/master/tests/modules/nf-core/bowtie2/build_test).
:::

### Migrating from pytest to nf-test

We recently decided to use nf-test instead of pytest for testing modules. This is because nf-test is more flexible and allows us to test modules in a more realistic way. You can find more information at [nf-test official docs](https://code.askimed.com/nf-test/) and [in this bytesize talk](https://nf-co.re/events/2022/bytesize_nftest).

> #### Philopsohy of nf-tests:

Reviewer: 🤦

> * Each module to now contain a `tests/` folder beside the `main.nf` containing the test files

Author: ```suggestion
* Each module contains a `tests/` folder beside the `main.nf` containing the test files
```

```groovy
setup {

            run("ABRICATE_RUN") {
                script "../../run/main.nf"
                process {
                    """
                    input[0] =  Channel.fromList([
                        tuple([ id:'test1', single_end:false ], // meta map
                            file(params.test_data['bacteroides_fragilis']['genome']['genome_fna_gz'], checkIfExists: true)),
                        tuple([ id:'test2', single_end:false ],
                            file(params.test_data['haemophilus_influenzae']['genome']['genome_fna_gz'], checkIfExists: true))
                    ])
                    """
                }
            }
        }
```

:::note
The setup method can run more than one process each enclosed in their `run` block
:::

- Then, the output of setup process/es can be provided as input in the `process` section of `when` block

```groovy=
input[0] = ABRICATE_RUN.out.report.collect{ meta, report -> report }.map{ report -> [[ id: 'test_summary'], report]}
```

> - Next, in the `then` block we can write our assertions that are used verify the test. A test can have multiple assertions but, we recommend enclosing all assertions in a `assertAll()` block as shown below:

Reviewer 2: ```suggestion
- Next, in the `then` block we can write our assertions that are used to verify the test. A test can have multiple assertions but, we recommend enclosing all assertions in a `assertAll()` block as shown below:
```

if you have a module named `build` this can conflict with some pytest internal behaviour. This results in no tests being run (i.e. receiving a message of `collected 0 items`). In this case rename the `tests/<module>/build` directory to `tests/<module>/build_test`, and update the corresponding `test.yml` accordingly. An example can be seen with the [`bowtie2/build` module tests](https://github.com/nf-core/modules/tree/master/tests/modules/nf-core/bowtie2/build_test).
:::

### Migrating from pytest to nf-test

We recently decided to use nf-test instead of pytest for testing modules. This is because nf-test is more flexible and allows us to test modules in a more realistic way. You can find more information at [nf-test official docs](https://code.askimed.com/nf-test/) and [in this bytesize talk](https://nf-co.re/events/2022/bytesize_nftest).

> #### Philosophy of nf-tests:

Author: ```suggestion
#### Philosophy of nf-tests
```

:::

### Migrating from pytest to nf-test

We recently decided to use nf-test instead of pytest for testing modules. This is because nf-test is more flexible and allows us to test modules in a more realistic way. You can find more information at [nf-test official docs](https://code.askimed.com/nf-test/) and [in this bytesize talk](https://nf-co.re/events/2022/bytesize_nftest).

#### Philopsohy of nf-tests:

- Each module to now contain a `tests/` folder beside the `main.nf` containing the test files
- Test files to come with a [snapshot](https://code.askimed.com/nf-test/docs/assertions/snapshots/) of module output channels

#### Steps for creating nf-test for a simple un-chained module

- Install or update to the latest version of [nf-test](https://code.askimed.com/nf-test/installation/)

```bash

> curl -fsSL https://code.askimed.com/install/nf-test | bash

Reviewer 3: Maybe a good idea to also add the conda install option here? 
```
conda install -c bioconda nf-test
```

The install step will copy the nf-test executable file in the current directory. If you want to use it globally, move the nf-test file to a directory accessible by your $PATH variable.
:::

- Or install via conda and make sure to update to the latest version

```bash

> conda install -c bioconda nf-test

Author: oh, didn't know about that. this should be the default imo, because the other install script just downloads it to the current folder and doesn't add it to the path

Reviewer: this doesn't get the latest version and im afraid ppl might forget to update....is that ok ?

Reviewer 3: I try to update the recipe whenever a new release happens. Also I think not everyone install nf-test everytime they create a test, right?

Author: having it on conda also makes CI easier. So with this we have more pressure to keep it updated there 🙂

Reviewer: thats right
